sqlmesh/core/table_diff.py

```diff
@@ -0,0 +1,195 @@
+
+
+class TableDiff:
+    """Calculates differences between tables, taking into account schema and row level differences."""
+
+    def __init__(
+        self,
+        adapter: EngineAdapter,
+        source: TableName,
+        target: TableName,
+        on: t.List[str] | exp.Condition,
+        where: t.Optional[str | exp.Condition] = None,
+        dialect: DialectType = None,
+        limit: int = 20,
+    ):
+        self.adapter = adapter
+        self.source = source
+        self.target = target
+        self.where = exp.condition(where, dialect=dialect) if where else None
+        self.limit = limit
+
+        if isinstance(on, list):
+            self.on: exp.Condition = exp.and_(
+                *(
+                    exp.column(c, "s").eq(exp.column(c, "t"))
+                    | (exp.column(c, "s").is_(exp.null()) & exp.column(c, "t").is_(exp.null()))
+                    for c in on
+                )
+            )
+        else:
+            self.on = on
+
+        self._source_schema: t.Optional[t.Dict[str, exp.DataType]] = None
+        self._target_schema: t.Optional[t.Dict[str, exp.DataType]] = None
+        self._summary_diff: t.Optional[SummaryDiff] = None
+
+    @property
+    def source_schema(self) -> t.Dict[str, exp.DataType]:
+        if self._source_schema is None:
+            self._source_schema = self.adapter.columns(self.source)
+        return self._source_schema
+
+    @property
+    def target_schema(self) -> t.Dict[str, exp.DataType]:
+        if self._target_schema is None:
+            self._target_schema = self.adapter.columns(self.target)
+        return self._target_schema
+
+    def schema_diff(self) -> SchemaDiff:
+        return SchemaDiff(
+            source=self.source,
+            target=self.target,
+            source_schema=self.source_schema,
+            target_schema=self.target_schema,
+        )
+
+    def summary_diff(self) -> SummaryDiff:
+        if self._summary_diff is None:
+            s_selects = {c: exp.column(c, "s").as_(f"s__{c}") for c in self.source_schema}
+            t_selects = {c: exp.column(c, "t").as_(f"t__{c}") for c in self.target_schema}
+
+            s_index = []
+            t_index = []
+
+            for col in self.on.find_all(exp.Column):
+                col.not_().is_(exp.Null())
+                if col.table == "s":
+                    s_index.append(col)
+                elif col.table == "t":
+                    t_index.append(col)
+
+            comparisons = [
+                exp.func("IF", exp.column(c, "s").eq(exp.column(c, "t")), 1, 0).as_(f"{c}_matches")
+                for c, t in self.source_schema.items()
+                if t == self.target_schema.get(c)
+            ]
+
+            query = (
+                exp.select(
+                    *s_selects.values(),
+                    *t_selects.values(),
+                    exp.func(
+                        "IF", exp.and_(*(c.not_().is_(exp.Null()) for c in s_index)), 1, 0
```

Comment:
Isn't it enough to have at least one non-null key to prove existence?

Reply:
no, you can have a null in a key